Rubberduck.CodeAnalysis/Inspections/Concrete/SetAssignmentWithIncompatibleObjectTypeInspection.cs

```diff
@@ -0,0 +1,229 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Inspections.Abstract;
+using Rubberduck.Inspections.Inspections.Extensions;
+using Rubberduck.Inspections.Results;
+using Rubberduck.Parsing;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Parsing.VBA.DeclarationCaching;
+using Rubberduck.Resources.Inspections;
+using Rubberduck.VBEditor;
+
+namespace Rubberduck.CodeAnalysis.Inspections.Concrete
+{
+    public class SetAssignmentWithIncompatibleObjectTypeInspection : InspectionBase
+    {
+        private readonly IDeclarationFinderProvider _declarationFinderProvider;
+
+        private const string UndeterminedValue = "Undetermined";
+
+        /// <summary>
+        /// Locates assignments to object variables for which the RHS does not have a compatible declared type. 
+        /// </summary>
+        /// <why>
+        /// The VBA compiler does not check whether different object types are compatible. Instead there is a runtime error whenever the types are incompatible.
+        /// </why>
+        /// <example>
+        /// <![CDATA[
+        /// IInterface:
+        ///
+        /// Public Sub DoSomething()
+        /// End Sub
+        ///
+        /// ------------------------------
+        /// Class1:
+        ///
+        ///'No Implements IInterface
+        /// 
+        /// Public Sub DoSomething()
+        /// End Sub
+        ///
+        /// ------------------------------
+        /// Module1:
+        /// 
+        /// Public Sub DoIt()
+        ///     Dim cls As Class1
+        ///     Dim intrfc As IInterface
+        ///
+        ///     Set cls = New Class1
+        ///     Set intrfc = cls 
+        /// End Sub
+        /// ]]>
+        /// </example>
+        /// <example>
+        /// <![CDATA[
+        /// IInterface:
+        ///
+        /// Public Sub DoSomething()
+        /// End Sub
+        ///
+        /// ------------------------------
+        /// Class1:
+        ///
+        /// Implements IInterface
+        /// 
+        /// Private Sub IInterface_DoSomething()
+        /// End Sub
+        ///
+        /// ------------------------------
+        /// Module1:
+        /// 
+        /// Public Sub DoIt()
+        ///     Dim cls As Class1
+        ///     Dim intrfc As IInterface
+        ///
+        ///     Set cls = New Class1
+        ///     Set intrfc = cls 
+        /// End Sub
+        /// ]]>
+        /// </example>
+        public SetAssignmentWithIncompatibleObjectTypeInspection(RubberduckParserState state)
+            : base(state)
+        {
+            _declarationFinderProvider = state;
```

Comment:
Shouldn't the default be set to error, rather than warning, given that any hits the inspection returns is most definitely a bug in the code?

Reply:
I guess, error makes sense.

However, note that there are situations in which the assignment will be legal. If you assign an object with one declared interface to a variable with another declared interface, the assignment is legal if the underlying object happens to implement both interfaces.

Reply:
That's why as alluded in the other comment below, such ambiguous cases would be a separate inspection with lower severity. The error inspection should only take cases where it is very clear that the code cannot run under any circumstances.

Reply:
I would like to differ. Only covering the cases for which we can prove that a runtime error will be raised will heavily limit the cases this inspection reports as an issue and make the implementation much more complicated because it would involve code path analysis. 

I think the VBE specifically does not check the declared object type beacuse under most circumstances it cannot know for certain which interfaces will be implemented on the object used in the assignment. 

It is always possible to split assignments into multiple steps to only use direct downcasts and upcasts.

Reply:
The severity is error now.